R/list-modify.R

```diff
@@ -37,72 +38,41 @@
 #' l <- list(new = 1, y = zap(), z = 5)
 #' str(list_modify(x, !!!l))
 list_modify <- function(.x, ...) {
-  list_recurse(.x, list2(...), function(x, y) y)
+  y <- dots_list(..., .named = NULL, .homonyms = "error")
+  list_recurse(.x, y, function(x, y) y)
 }
 #' @export
 #' @rdname list_modify
 list_merge <- function(.x, ...) {
-  list_recurse(.x, list2(...), c)
+  y <- dots_list(..., .named = NULL, .homonyms = "error")
+  list_recurse(.x, y, c)
 }
 
-list_recurse <- function(x, y, base_case) {
+list_recurse <- function(x, y, base_f) {
   stopifnot(is.list(x), is.list(y))
 
-  if (is_empty(x)) {
-    return(y)
-  }
-  if (is_empty(y)) {
-    return(x)
-  }
-
-  y_names <- names(y)
-  if (!is_null(y_names) && !is_named(y)) {
+  if (!is_null(names(y)) && !is_named(y)) {
     abort("`...` arguments must be either all named, or all unnamed")
   }
 
-  # N.B. is_list(zap()) is TRUE.
-  if (is_null(y_names)) {
-    for (i in rev(seq_along(y))) {
-      if (i <= length(x) && is_list(x[[i]]) && is_list(y[[i]]) && !is_zap(y[[i]])) {
-        x[[i]] <- list_recurse(x[[i]], y[[i]], base_case)
-      } else {
-        x[[i]] <- maybe_zap(base_case(x[[i]], y[[i]]))
-      }
-    }
-  } else {
-    for (i in seq_along(y_names)) {
-      nm <- y_names[[i]]
-      if (has_name(x, nm) && is_list(x[[nm]]) && is_list(y[[i]]) && !is_zap(y[[i]])) {
-        x[[nm]] <- list_recurse(x[[nm]], y[[i]], base_case)
-      } else {
-        x[[nm]] <- maybe_zap(base_case(x[[nm]], y[[i]]))
-      }
-    }
-  }
+  idx <- names(y) %||% rev(seq_along(y))
 
-  x
-}
+  for (i in idx) {
+    x_i <- pluck(x, i)
+    y_i <- pluck(y, i)
 
-maybe_zap <- function(x) {
-  if (is_zap(x)) {
-    return(NULL)
-  }
-  if (!is_null(x)) {
-    return(x)
+    if (is_zap(y_i)) {
+      x[[i]] <- NULL
+    } else if (vec_is_list(x_i) && vec_is_list(y_i)) {
+      list_slice2(x, i) <- list_recurse(x_i, y_i, base_f)
+    } else {
```

Comment:
I think one thing that has always bothered me in these functions is this very flexible base case, i.e. that it accepts mismatches between lists and atoms. Should we require that the list structure of the lhs and rhs match, and fail otherwise?

Reply:
Yeah, that's a good idea. I'll explore.

Reply:
I _think_ this is ok — it lets you override vectors with lists and vice versa:

``` r
library(purrr)

str(list_modify(list(x = list(1, 2)), x = 2))
#> List of 1
#>  $ x: num 2
str(list_modify(list(x = 2), x = list(1, 2)))
#> List of 1
#>  $ x:List of 2
#>   ..$ : num 1
#>   ..$ : num 2
```

<sup>Created on 2022-09-07 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>